File: Telemedicine-and-Doctor-Consultation-Platform-main/backend/routes/inventory.js
const router = require('express').Router();
const auth   = require('../middleware/auth');
const db     = require('../config/db');

// GET /api/inventory
router.get('/', auth, async (req, res) => {
  try {
    const { low } = req.query;
    let query = `
      SELECT i.*, m.name, m.price 
      FROM inventory i 
      JOIN medicines m ON i.sku = m.sku
    `;
    
    if (low === 'true') {
      query += ' WHERE i.qty < i.min_qty';
    }
    
    const [rows] = await db.query(query);
    res.json({ count: rows.length, data: rows });
  } catch (err) {
    res.status(500).json({ error: 'Database error', message: err.message });
  }
});

// GET /api/inventory/:sku
router.get('/:sku', auth, async (req, res) => {
  try {
    const [rows] = await db.query(`
      SELECT i.*, m.name, m.price, m.brand 
      FROM inventory i 
      JOIN medicines m ON i.sku = m.sku 
      WHERE i.sku = ?
    `, [req.params.sku]);
    
    if (!rows.length) return res.status(404).json({ error: 'Item not found' });
    res.json(rows[0]);
  } catch (err) {
    res.status(500).json({ error: 'Database error', message: err.message });
  }
});

// POST /api/inventory
router.post('/', auth, async (req, res) => {
  const connection = await db.getConnection();
  try {
    const { name, qty, minQty, price, expiry } = req.body;
    if (!name || qty === undefined) return res.status(400).json({ error: 'name and qty required' });
    
    await connection.beginTransaction();

    const [countRows] = await connection.query('SELECT COUNT(*) as count FROM medicines');
    const newSku = `MED-${String(countRows[0].count + 1).padStart(3,'0')}`;
    
    // Insert into Medicines
    await connection.query(
      `INSERT INTO medicines (sku, name, price) VALUES (?, ?, ?)`,
      [newSku, name, price || 0]
    );

    // Insert into Inventory
    await connection.query(
      `INSERT INTO inventory (sku, qty, min_qty, expiry_date) VALUES (?, ?, ?, ?)`,
      [newSku, qty, minQty || 20, expiry || null]
    );

    await connection.commit();
    res.status(201).json({ sku: newSku, name, qty, minQty, price, expiry });
  } catch (err) {
    await connection.rollback();
    res.status(500).json({ error: 'Database error', message: err.message });
  } finally {
    connection.release();
  }
});

// PATCH /api/inventory/:sku — restock
router.patch('/:sku', auth, async (req, res) => {
  try {
    const updates = [];
    const params = [];
    if (req.body.qty !== undefined) { updates.push('qty = ?'); params.push(req.body.qty); }
    if (req.body.minQty !== undefined) { updates.push('min_qty = ?'); params.push(req.body.minQty); }
    
    if (updates.length) {
      params.push(req.params.sku);
      await db.query(`UPDATE inventory SET ${updates.join(', ')} WHERE sku = ?`, params);
    }
    
    const [rows] = await db.query('SELECT * FROM inventory WHERE sku = ?', [req.params.sku]);
    if (!rows.length) return res.status(404).json({ error: 'Item not found' });
    
    res.json(rows[0]);
  } catch (err) {
    res.status(500).json({ error: 'Database error', message: err.message });
  }
});

// GET /api/inventory/alerts/low-stock
router.get('/alerts/low-stock', auth, async (req, res) => {
  try {
    const [rows] = await db.query(`
      SELECT i.*, m.name 
      FROM inventory i 
      JOIN medicines m ON i.sku = m.sku 
      WHERE i.qty < i.min_qty
    `);
    res.json({ count: rows.length, items: rows });
  } catch (err) {
    res.status(500).json({ error: 'Database error', message: err.message });
  }
});

module.exports = router;

File: backend/routes/inventory.py
from flask import Blueprint, request, jsonify
from config import get_db_connection
from middleware.auth import token_required

bp = Blueprint('inventory', __name__, url_prefix='/api/inventory')

@bp.route('/', methods=['GET'])
@token_required
def get_inventory():
    try:
        low = request.args.get('low')
        
        query = '''SELECT i.*, m.name, m.price 
                   FROM inventory i 
                   JOIN medicines m ON i.sku = m.sku'''
        
        if low == 'true':
            query += ' WHERE i.qty < i.min_qty'
        
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute(query)
        rows = cursor.fetchall()
        cursor.close()
        conn.close()
        
        return jsonify({'count': len(rows), 'data': rows}), 200
    
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500

@bp.route('/<sku>', methods=['GET'])
@token_required
def get_inventory_item(sku):
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute(
            '''SELECT i.*, m.name, m.price, m.brand 
               FROM inventory i 
               JOIN medicines m ON i.sku = m.sku 
               WHERE i.sku = %s''',
            (sku,)
        )
        item = cursor.fetchone()
        cursor.close()
        conn.close()
        
        if not item:
            return jsonify({'error': 'Item not found'}), 404
        
        return jsonify(item), 200
    
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500

@bp.route('/', methods=['POST'])
@token_required
def create_inventory():
    try:
        data = request.get_json()
        name = data.get('name')
        qty = data.get('qty')
        min_qty = data.get('minQty')
        price = data.get('price')
        expiry = data.get('expiry')
        
        if not name or qty is None:
            return jsonify({'error': 'name and qty required'}), 400
        
        conn = get_db_connection()
        cursor = conn.cursor()
        
        try:
            # Get count for SKU
            cursor.execute('SELECT COUNT(*) as count FROM medicines')
            result = cursor.fetchone()
            count = result['count'] + 1
            new_sku = f'MED-{str(count).zfill(3)}'
            
            # Insert into medicines
            cursor.execute(
                'INSERT INTO medicines (sku, name, price) VALUES (%s, %s, %s)',
                (new_sku, name, price or 0)
            )
            
            # Insert into inventory
            cursor.execute(
                'INSERT INTO inventory (sku, qty, min_qty, expiry_date) VALUES (%s, %s, %s, %s)',
                (new_sku, qty, min_qty or 20, expiry)
            )
            
            conn.commit()
            
            return jsonify({
                'sku': new_sku,
                'name': name,
                'qty': qty,
                'minQty': min_qty or 20,
                'price': price,
                'expiry': expiry
            }), 201
        
        except Exception as err:
            conn.rollback()
            raise err
        finally:
            cursor.close()
            conn.close()
    
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500

@bp.route('/<sku>', methods=['PATCH'])
@token_required
def update_inventory(sku):
    try:
        data = request.get_json()
        updates = []
        params = []
        
        if 'qty' in data and data['qty'] is not None:
            updates.append('qty = %s')
            params.append(data['qty'])
        if 'minQty' in data and data['minQty'] is not None:
            updates.append('min_qty = %s')
            params.append(data['minQty'])
        
        if updates:
            params.append(sku)
            
            conn = get_db_connection()
            cursor = conn.cursor()
            cursor.execute(f"UPDATE inventory SET {', '.join(updates)} WHERE sku = %s", params)
            conn.commit()
        else:
            conn = get_db_connection()
            cursor = conn.cursor()
        
        cursor.execute('SELECT * FROM inventory WHERE sku = %s', (sku,))
        item = cursor.fetchone()
        cursor.close()
        conn.close()
        
        if not item:
            return jsonify({'error': 'Item not found'}), 404
        
        return jsonify(item), 200
    
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500

@bp.route('/alerts/low-stock', methods=['GET'])
@token_required
def get_low_stock():
    try:
        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute(
            '''SELECT i.*, m.name 
               FROM inventory i 
               JOIN medicines m ON i.sku = m.sku 
               WHERE i.qty < i.min_qty'''
        )
        rows = cursor.fetchall()
        cursor.close()
        conn.close()
        
        return jsonify({'count': len(rows), 'items': rows}), 200
    
    except Exception as err:
        return jsonify({'error': 'Database error', 'message': str(err)}), 500
